rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -98,17 +100,74 @@ def __str__(self) -> str:
+    def __init__(self) -> None:
+        self._acquisitions = deque()
+
+    @abstractmethod
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        """The core method. Should be implemented by subclasses.
+            This method takes a new acquisition and decides where to insert it
+            This method determines the PnL order
+        """
+        ...
+
+    def __next__(self) -> Optional[AssetAcquisitionEvent]:
+        """Iteration method over acquisition events. Returns current front event
+        May raise:
+            StopIterationError - when we have iterated over all events
+        """
+        try:
+            return self._acquisitions[0]
+        except IndexError as e:
+            raise StopIteration from e
+
+    def report_result(self, used_amount: FVal) -> None:
+        """This function should be used to report results of the
+        currently processed event (received from __next__)
+        The current event's remaining_amount will be decreased by used_amount
+        If event's remaining_amount will become ZERO, the event will be deleted
+        May raise:
+            ValueError if used_amount is incorrect
+            IndexError if the method was called when acquisitions were empty
+        """
+        if used_amount < ZERO or used_amount > self._acquisitions[0].remaining_amount:
+            raise ValueError(f'Used amount should be in the interval [0, {self._acquisitions[0].remaining_amount}]')  # noqa: E501
+
+        self._acquisitions[0].remaining_amount -= used_amount
+        if self._acquisitions[0].remaining_amount == ZERO:
+            self._acquisitions.popleft()
+
+
+class FIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in FIFO (first-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        self._acquisitions.append(acquisition)
+
+
+class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in LIFO (last-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        self._acquisitions.appendleft(acquisition)
+
+
 class CostBasisEvents:
-    used_acquisitions: List[AssetAcquisitionEvent] = field(init=False)
-    acquisitions: List[AssetAcquisitionEvent] = field(init=False)
-    spends: List[AssetSpendEvent] = field(init=False)
+    used_acquisitions: List[AssetAcquisitionEvent]
+    acquisitions: BaseAcquisitionsOrder
+    spends: List[AssetSpendEvent]
 
-    def __post_init__(self) -> None:
+    def __init__(self, pnl_order: PnlOrder) -> None:
         """Using this since can't use mutable default arguments"""
-        self.used_acquisitions = []
-        self.acquisitions = []
+        if pnl_order == PnlOrder.FIFO:
+            self.acquisitions = FIFOAcquisitionsOrder()
+        elif pnl_order == PnlOrder.LIFO:
+            self.acquisitions = LIFOAcquisitionsOrder()
+        else:
+            raise ValueError(f'Unknown pnl order: {pnl_order}')
```

Comment:
Never ever use `ValueError`. Both here and above you seem to raise `ValueError` which is the most generic exception. In our code we always raise our own exceptions and whenever we raise an exception we make sure to catch it somewhere. Otherwise it's a bug and at some point will blow up on a user's face.

For example here this is not an error that can happen. This is a programing error that **should never happen**. So this is `AssertionError`. 

Keep in mind assertions can be optimized and removed at runtime if `assert condition, msg` is used. But `raise AssertionError` can not.

Still I guess here is fine to use `AssertionError`